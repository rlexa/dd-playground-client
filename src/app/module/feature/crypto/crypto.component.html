<div class="list-h wrap">
  <mat-card appearance="outlined" class="childhv">
    <mat-card-title>Actions: API</mat-card-title>
    <mat-card-content>
      <div class="list-v">
        <h1 class="color-warn">CORS/PROXY KILLS API.</h1>

        <mat-list class="mat-list-auto-height">
          <h3 mat-subheader>Ping</h3>
          <mat-list-item>
            <div class="bar-h">
              <button mat-button (click)="wsBlockchainPing()">Blockchain WS</button>
              <button mat-button (click)="wsBlockCypherPing()">BlockCypher WS</button>
            </div>
          </mat-list-item>

          <h3 mat-subheader>Address</h3>
          <mat-list-item>
            <div class="bar-h">
              <mat-form-field class="childh value-numeric">
                <input
                  matInput
                  [ngModel]="bitcoinAddress$ | async"
                  (ngModelChange)="bitcoinAddress$.next($event)"
                  placeholder="Bitcoin Address"
                />
              </mat-form-field>
              <div class="list-v">
                <button mat-button (click)="apiBlockchainBitcoinAddress()">Blockchain API</button>
                <button mat-button (click)="wsBlockchainAddressLatestTransaction()">Blockchain WS</button>
              </div>
              <div class="list-v">
                <button mat-button (click)="apiBlockCypherBitcoinAddress()">BlockCypher API</button>
                <button mat-button (click)="wsBlockCypherAddressConfirmedTx()">BlockCypher WS</button>
              </div>
            </div>
          </mat-list-item>

          <h3 mat-subheader>Stats</h3>
          <mat-list-item>
            <div class="bar-h">
              <div class="list-v">
                <button mat-button (click)="apiWebBitcoinStats()">WebBitcoin API</button>
              </div>
            </div>
          </mat-list-item>
        </mat-list>
      </div>
    </mat-card-content>
  </mat-card>

  <mat-card appearance="outlined" class="childhv">
    <mat-card-title>Actions: Output</mat-card-title>
    <mat-card-content>
      <app-simple-view [data]="anyData$ | async" [isWidthConstrained]="false" [isDense]="true" />
    </mat-card-content>
  </mat-card>
</div>
