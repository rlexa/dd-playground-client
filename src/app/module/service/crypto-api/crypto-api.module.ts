import { NgModule } from '@angular/core';
import { imports } from './imports';

@NgModule({ imports })
export class CryptoApiModule { }
