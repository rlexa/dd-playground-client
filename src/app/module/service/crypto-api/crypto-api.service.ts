import {HttpClient} from '@angular/common/http';
import {Injectable} from '@angular/core';
import {cleanParams, getJsonHeaders} from 'src/app/util';
import {of} from 'rxjs';
import {map, switchMap, take, tap} from 'rxjs/operators';
import {WebSocketSubject} from 'rxjs/webSocket';

class BlockchainWebSocketSubject extends WebSocketSubject<object> {
  ping = () => this.next({op: 'ping'});

  addressLatestTransaction = (address: string) => {
    this.next({op: 'addr_sub', addr: address});
    this.next({op: 'ping_tx'});
    this.next({op: 'addr_unsub', addr: address});
  };
}

class BlockCypherWebSocketSubject extends WebSocketSubject<object> {
  ping = () => this.next({event: 'ping'});

  addressConfirmedTx = (address: string) => this.next({event: 'confirmed-tx', address});
}

export interface RequestSkipTake {
  skip?: number;
  take?: number;
}
export interface RequestAddressItems extends RequestSkipTake {
  address: string;
}

@Injectable({providedIn: 'root'})
export class CryptoApiService {
  constructor(private readonly http: HttpClient) {}

  private readonly API_BLOCKCHAIN = 'https://blockchain.info';
  private readonly API_BLOCKCYPHER = 'https://api.blockcypher.com/v1';
  private readonly API_WEBBITCOIN = 'http://webbtc.com';
  private readonly WS_BLOCKCHAIN = 'wss://ws.blockchain.info/inv';
  private readonly WS_BLOCKCYPHER = 'wss://socket.blockcypher.com/v1/btc/main';

  wsBlockchainPing$ = () =>
    of(this.wsBlockchain()).pipe(
      tap(_ => _.ping()),
      switchMap(_ => _),
      take(1),
    );
  wsBlockchainAddressLatestTransaction$ = (address: string) =>
    of(this.wsBlockchain()).pipe(
      tap(_ => _.addressLatestTransaction(address)),
      switchMap(_ => _),
      take(1),
    );

  wsBlockCypherPing$ = () =>
    of(this.wsBlockCypher()).pipe(
      tap(_ => _.ping()),
      switchMap(_ => _),
      take(1),
    );
  wsBlockCypherAddressConfirmedTx$ = (address: string) =>
    of(this.wsBlockCypher()).pipe(
      tap(_ => _.addressConfirmedTx(address)),
      switchMap(_ => _),
      take(1),
    );

  blockchainSingleAddress$ = (req: RequestAddressItems) =>
    this.httpGetRest$({
      api: this.API_BLOCKCHAIN,
      url: `/rawaddr/${req.address || 'null'}`,
      params: {limit: req.take, offset: req.skip, cors: true},
    }).pipe(map(_ => _));

  blockcypherSingleAddress$ = (req: RequestAddressItems) =>
    this.httpGetRest$({
      api: this.API_BLOCKCYPHER,
      url: `/btc/main/addrs/${req.address || 'null'}`,
      params: {limit: req.take, offset: req.skip},
    }).pipe(map(_ => _));

  webbitcoinStats$ = () => this.httpGetRest$({api: this.API_WEBBITCOIN, url: `/stats.json`}).pipe(map(_ => _));

  private httpGetRest$ = (val: {api: string; url: string; params?: any}) =>
    this.http.get(val.api + val.url, {headers: getJsonHeaders(), params: cleanParams(val.params)});

  private wsBlockchain = () => new BlockchainWebSocketSubject(this.WS_BLOCKCHAIN);
  private wsBlockCypher = () => new BlockCypherWebSocketSubject(this.WS_BLOCKCYPHER);
}
