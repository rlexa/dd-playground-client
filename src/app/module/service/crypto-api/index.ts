export * from './crypto-api.module';
export * from './crypto-api.service';
