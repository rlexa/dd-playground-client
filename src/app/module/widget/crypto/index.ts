export * from './crypto.component';
export * from './crypto.module';
