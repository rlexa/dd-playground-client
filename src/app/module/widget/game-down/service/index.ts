export * from './game-down.service';
