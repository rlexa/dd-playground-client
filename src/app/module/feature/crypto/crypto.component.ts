import {CommonModule} from '@angular/common';
import {ChangeDetectionStrategy, Component, OnDestroy, inject} from '@angular/core';
import {FormsModule} from '@angular/forms';
import {MatButtonModule} from '@angular/material/button';
import {MatCardModule} from '@angular/material/card';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatInputModule} from '@angular/material/input';
import {MatListModule} from '@angular/material/list';
import {RxCleanup, rxNext_} from 'dd-rxjs';
import {BehaviorSubject, Subject, of} from 'rxjs';
import {switchMap} from 'rxjs/operators';
import {CryptoApiService} from '../../service/crypto-api';
import {SimpleViewComponent} from '../../widget/simple-view';

@Component({
  selector: 'app-crypto',
  templateUrl: './crypto.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: true,
  imports: [
    CommonModule,
    FormsModule,
    MatButtonModule,
    MatCardModule,
    MatFormFieldModule,
    MatInputModule,
    MatListModule,
    SimpleViewComponent,
  ],
})
export class CryptoComponent implements OnDestroy {
  private readonly api = inject(CryptoApiService);

  @RxCleanup() readonly bitcoinAddress$ = new BehaviorSubject('1BoatSLRHtKNngkdXEeobR76b53LETtpyT');
  @RxCleanup() readonly anyData$ = new Subject();

  destroy() {}
  ngOnDestroy() {
    this.destroy();
  }

  apiBlockchainBitcoinAddress = () =>
    of(this.bitcoinAddress$.value || null)
      .pipe(switchMap((_) => this.api.blockchainSingleAddress$({address: _})))
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));

  apiBlockCypherBitcoinAddress = () =>
    of(this.bitcoinAddress$.value || null)
      .pipe(switchMap((_) => this.api.blockcypherSingleAddress$({address: _})))
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));

  apiWebBitcoinStats = () =>
    of(this.bitcoinAddress$.value || null)
      .pipe(switchMap((_) => this.api.webbitcoinStats$()))
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));

  wsBlockchainPing = () =>
    of(null)
      .pipe(switchMap((_) => this.api.wsBlockchainPing$()))
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));

  wsBlockchainAddressLatestTransaction = () =>
    of(this.bitcoinAddress$.value || null)
      .pipe(switchMap((_) => this.api.wsBlockchainAddressLatestTransaction$(_)))
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));

  wsBlockCypherPing = () =>
    of(null)
      .pipe(switchMap((_) => this.api.wsBlockCypherPing$()))
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));

  wsBlockCypherAddressConfirmedTx = () =>
    of(this.bitcoinAddress$.value || null)
      .pipe(switchMap((_) => this.api.wsBlockCypherAddressConfirmedTx$(_)))
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));
}
