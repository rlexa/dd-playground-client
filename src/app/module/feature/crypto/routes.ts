import {Route} from '@angular/router';
import {RouteBlockchain, RouteRoot, RouteWild} from 'src/app/routing';
import {NavigationBarItem, NavigationBarItemsData} from '../../widget/navigation-bar';
import {NavigationContentComponent} from '../../widget/navigation-content';

const routeNavs: Record<string, NavigationBarItem> = {
  [RouteBlockchain]: {icon: 'items1', route: RouteBlockchain, label: 'Blockchain'},
};

const data: NavigationBarItemsData = {
  navigationBarItems: Object.values(routeNavs),
};

export default [
  {
    path: RouteRoot,
    component: NavigationContentComponent,
    data,
    children: [
      {path: RouteBlockchain, loadComponent: () => import('./crypto.component').then((m) => m.CryptoComponent)},
      {path: RouteRoot, redirectTo: RouteBlockchain, pathMatch: 'full'},
      {path: RouteWild, redirectTo: RouteBlockchain},
    ],
  },
] as Route[];
