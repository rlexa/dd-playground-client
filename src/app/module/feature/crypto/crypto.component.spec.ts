import {ComponentFixture} from '@angular/core/testing';
import {MockBuilder, MockRender} from 'ng-mocks';
import {mockAll} from 'src/app/test';
import {CryptoApiService} from '../../service/crypto-api';
import {CryptoComponent} from './crypto.component';

describe('CryptoComponent', () => {
  let fixture: ComponentFixture<CryptoComponent>;

  beforeEach(() => MockBuilder(CryptoComponent).provide(mockAll(CryptoApiService)));

  beforeEach(() => {
    fixture = MockRender(CryptoComponent);
    fixture.detectChanges();
  });

  test('renders', () => expect(fixture).toMatchSnapshot());
});
