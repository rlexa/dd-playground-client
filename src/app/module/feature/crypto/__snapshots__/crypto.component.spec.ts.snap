// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoComponent renders 1`] = `
<mock-render>
  <app-crypto>
    <div
      ddflexbox="hyscsw"
      ng-reflect-dd-flexbox="hyscsw"
    >
      <mat-card
        appearance="outlined"
        class="childhv"
        ng-reflect-appearance="outlined"
      >
        <mat-card-title>
          Actions: API
        </mat-card-title>
        <mat-card-content>
          <div
            ddflexbox="vys"
            ng-reflect-dd-flexbox="vys"
          >
            <h1
              class="color-warn"
            >
              CORS/PROXY KILLS API.
            </h1>
            <mat-list
              class="mat-list-auto-height"
            >
              <h3
                mat-subheader=""
              >
                Ping
              </h3>
              <mat-list-item>
                <div
                  ddflexbox="hyc"
                  ng-reflect-dd-flexbox="hyc"
                >
                  <button
                    mat-button=""
                  >
                    Blockchain WS
                  </button>
                  <button
                    mat-button=""
                  >
                    BlockCypher WS
                  </button>
                </div>
              </mat-list-item>
              <h3
                mat-subheader=""
              >
                Address
              </h3>
              <mat-list-item>
                <div
                  ddflexbox="hyc"
                  ng-reflect-dd-flexbox="hyc"
                >
                  <mat-form-field
                    class="childh value-numeric"
                  >
                    <input
                      matinput=""
                      ng-reflect-model="1BoatSLRHtKNngkdXEeobR76b53LET"
                      ng-reflect-placeholder="Bitcoin Address"
                      placeholder="Bitcoin Address"
                    />
                  </mat-form-field>
                  <div
                    ddflexbox="vys"
                    ng-reflect-dd-flexbox="vys"
                  >
                    <button
                      mat-button=""
                    >
                      Blockchain API
                    </button>
                    <button
                      mat-button=""
                    >
                      Blockchain WS
                    </button>
                  </div>
                  <div
                    ddflexbox="vys"
                    ng-reflect-dd-flexbox="vys"
                  >
                    <button
                      mat-button=""
                    >
                      BlockCypher API
                    </button>
                    <button
                      mat-button=""
                    >
                      BlockCypher WS
                    </button>
                  </div>
                </div>
              </mat-list-item>
              <h3
                mat-subheader=""
              >
                Stats
              </h3>
              <mat-list-item>
                <div
                  ddflexbox="hyc"
                  ng-reflect-dd-flexbox="hyc"
                >
                  <div
                    ddflexbox="vys"
                    ng-reflect-dd-flexbox="vys"
                  >
                    <button
                      mat-button=""
                    >
                      WebBitcoin API
                    </button>
                  </div>
                </div>
              </mat-list-item>
            </mat-list>
          </div>
        </mat-card-content>
      </mat-card>
      <mat-card
        appearance="outlined"
        class="childhv"
        ng-reflect-appearance="outlined"
      >
        <mat-card-title>
          Actions: Output
        </mat-card-title>
        <mat-card-content>
          <app-simple-view
            ng-reflect-is-dense="true"
            ng-reflect-is-width-constrained="false"
          />
        </mat-card-content>
      </mat-card>
    </div>
  </app-crypto>
</mock-render>
`;
